doc/developer/design/20231002_cluster_vision.md

```diff
@@ -0,0 +1,124 @@
+* Make product prioritization decisions around cluster work
+* Make folks more comfortable accepting intermediate states that aren't ideal in service of a greater goal
+* Come up with a narrative for customers on what to expect around cluster management
+
+Epic: https://github.com/MaterializeInc/materialize/issues/22120
+
+## Success Criteria
+Primarily, a merged design doc that is reviewed and approved by EPD leadership,
+and is socialized to GTM.
+
+Secondarily, a roadmap for cluster work for the next quarter.
+
+Qualitatively, positive feedback from EPD leadership and GTM folks that they
+have clarity on the vision and roadmap, and the reasoning behind those decisions.
+
+## Out of Scope
+Designing the actual cluster API changes themselves, or proposing implementation details.
+
+## Solution Proposal
+The objectives we are striving for with the cluster UX:
+* Easy to use and manage
+* Maximize resource efficiency/minimize unused resource cost
+* Enable fault tolerance/use-case isolation
+
+### Declarative vs Imperative
+We should move toward a declarative API for managing clusters, where:
+
+Declarative is like `CREATE CLUSTER` with managed replicas and \
+Imperative is like `CREATE/DROP CLUSTER REPLICA`.
+
+This means deprecating manual cluster replica management. \
+We believe this is easier to use and manage.
+
+We can classify actions that users take in managing clusters into two categories:
+_development workflows_ and _production workflow_.
+
+#### Development workflows
+In a development workflow, the underlying set of objects being configured are not being used yet
+in a production system, and the user is rapidly changing things.\
+In this workflow, downtime is acceptable.\
+A command like `ALTER <object> ... SET CLUSTER` (moving an object between clusters) would fall
+under this category.
+
+For development workflows, since downtime is acceptable, the primary work items is to
+**expose rehydration status**.\
+Users need an easy way to detect that rehydration is complete and they can resume querying against
+the object.
+
+#### Production workflows
+In a production workflow, the underlying set of objects are actively depended on by a production
+system.\
+In this workflow, downtime is not acceptable.\
+A command like `ALTER CLUSTER ... SET (SIZE = <>)` (resizing a cluster) would fall under this
+category.
+
+If a user wants to do a development workflow on a production system, they must use **blue/green
+deployments**. For example, if the user wants to move an object between clusters, they must use
+blue/green to set up another version of the object/cluster and cutover the production system
+to it once the object is rehydrated and ready.\
+Again, for this workflow, exposing hydration status is the primary work item.
+
+For production workflows, like resizing an active cluster, blue/green is an acceptable intermediate
+solution, but is an overkill amount of work for such a simple action.
```

Comment:
This text is confusing to me! What would it mean to blue/green a resized cluster? Like, the act of resizing would amount to creating a new cluster, with different resources behind it, and cutting over from one to the other? It is hard for me to understand this in the context of a blue/green implementation that e.g. rebuild and renames things, where downstream dependents are left confused. Would we drop sinks when we do this, for example?

Reply:
Agreed - Frank and I discussed offline and blue/green is too burdensome for scaling use-cases, and particularly with the versions of blue/green that we will realistically be building now which will be manually-controlled.